dashboard/app/models/user.rb

```diff
@@ -1243,11 +1243,25 @@ def reset_secrets
   end
 
   def generate_secret_picture
-    self.secret_picture = SecretPicture.random
+    new_secret_picture = SecretPicture.random
+
+    # regenerate if random picture is same as user's current secret picture
+    if new_secret_picture == secret_picture
+      generate_secret_picture
+    else
+      self.secret_picture = new_secret_picture
+    end
   end
 
   def generate_secret_words
-    self.secret_words = [SecretWord.random.word, SecretWord.random.word].join(" ")
+    new_secret_words = [SecretWord.random.word, SecretWord.random.word].join(" ")
+
+    # regenerate if random words are same as user's current secret words
+    if new_secret_words == secret_words
+      generate_secret_words
+    else
+      self.secret_words = new_secret_words
+    end
   end
```

Comment:
It would be nice to have an upper bound on the number of attempts here, and also to solve this with iteration instead of recursion since we [aren't guaranteed tail-call optimization](https://stackoverflow.com/questions/824562/does-ruby-perform-tail-call-optimization).

That might seem pedantic, but we live in a world where all edge cases happen. :grin:  Check out [the CodeGeneration module](https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/code_generation.rb) - there might be something there we can reuse.

Reply:
Very interesting - thanks for adding these links!

Reply:
that's a great point! a few thoughts after talking to @caleybrock in person:
- this could have a low number of `MAX_RETRIES` since it's less likely to fail (3? 5?)
- even if we hit the max number of retries, should we send an error to the user, which feels weird to do from the model, or just set the new secret and move on?

Reply:
:expressionless: I'd set the new secret and move on.  It's easy enough for the user to manually retry if they really want to.

Alternatively, I think it'd be easy enough to modify the `random` methods on the `SecretPicture` and `SecretWord` models so that they construct a query that excludes the existing secrets - all the necessary indices are already in place and these are small sets.

Reply:
Another way to deflake this test would be to take the randomness out of it:

https://github.com/code-dot-org/code-dot-org/blob/6d2b400da0378c81f9197c49d2bbe5e4b463b12a/dashboard/test/models/sections/section_test.rb#L57-L58

Reply:
And, just for completeness, [TCO is possible in Ruby](http://nithinbekal.com/posts/ruby-tco/) but reading that blog post it doesn't seem _worth_ it...

Reply:
i considered removing the randomness from the test, but thought it was confusing for a user to reset a secret and possibly receive the same secret in response. obviously not the worst that could happen, but 🤷‍♀️ 

i'll add a retry count for safety! thanks for your help, i appreciate the feedback

Reply:
and for teaching me about TCO!